Concurrently flush records in goroutines with concurrency level

zackwine opened the pull request:
Issue #, if available:
Related to: fluent/fluent-bit#2159
Updated version of this PR:  https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/pull/28

Description of changes:

These changes expand on the concurrent flush `goroutines` branch from @PettitWesley by adding a few features and fixing several issues:

- Makes the concurrent flush optional via `concurrency` option
- Adds a configurable limit on the number of concurrent goroutines: `concurrency`
- Improves retry logic to ensure records are not dropped when backpressure is received from kinesis
- Fixes an issue where records are sent more than once if backpressure is received from kinesis (only fixed if concurrency > 0)

From @PettitWesley related to this feature:

Previously, this plugin was synchronous. Since Fluent Bit is single-threaded, this decreases performance. Native C plugins in Fluent Bit core can take advantage of its event loop to asynchronously make requests. Go plugins are not as fancy. With this change, the plugin returns control to Fluent Bit immediately and sends data in a goroutine. I've previously verified that the goroutines are proper separate threads and continue running even after control returns to the Fluent Bit C code.

The downside to this change is that plugins will not accurately report errors or retries to Fluent Bit. I'm considering making this change an optional mode that is turned off by default and users can opt into if they want.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

PettitWesley committed:
Remove exponential backoff code

PettitWesley committed:
experimental: return immediately and send data in a goroutine

PettitWesley committed:
Add retries to goroutine/concurrent mode

PettitWesley committed:
flush on exit does not make sense in goroutine/concurrent mode

PettitWesley committed:
tmp

PettitWesley committed:
its not working...

PettitWesley committed:
Hmmm...

PettitWesley committed:
Need to use a pointer to a slice

PettitWesley committed:
Fix bug from backoff code change

PettitWesley committed:
Concurrency fix: process incoming C data structures before returning

PettitWesley committed:
Fix record length bug

PettitWesley committed:
Debugging

PettitWesley committed:
Debugging...

PettitWesley committed:
Debugging...

PettitWesley committed:
Oops

PettitWesley committed:
hmmm

PettitWesley committed:
debugging

PettitWesley committed:
What is going on

PettitWesley committed:
makes no sense at all

PettitWesley committed:
WAT THE

PettitWesley committed:
cats

PettitWesley committed:
Call AddRecord before spawning goroutine

PettitWesley committed:
Remove log message used for debug

PettitWesley committed:
Remove invalid partition key field from output plugin- todo: fix later

PettitWesley committed:
Fix bug in error handling logic

PettitWesley committed:
Remove unneeded info statements

zackwine committed:
Fix retry logic for concurrent flushes.

Fixes duplicate logs when kinesis limits are exceeded.

Reduces the likelihood of dropped logs when kinesis provides backpressure (improvements in retries).

Adds log messages to indicate when and how many records are dropped due to retry timeouts.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Fix retry logic for concurrent flushes.

Fixes duplicate logs when kinesis limits are exceeded.

Reduces the likelihood of dropped logs when kinesis provides backpressure (improvements in retries).

Adds log messages to indicate when and how many records are dropped due to retry timeouts.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Rename backoff to currentRetries for clarity.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Rename backoff to currentRetries for clarity.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Make limitsExceeded var to ensure message is valid.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Address code review.  Make limitsExceeded var to ensure message is valid.

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine committed:
Add concurrency option, and make concurrent goroutine flushes optional.

zackwine committed:
Merge branch 'goroutines' into goroutines-optional

zackwine committed:
Address code review.  Update docs, and make retries configurable.

zackwine committed:
Merge master

PettitWesley commented:
@zackwine Lol I love how you kept all of my original commit messages... 

🤦🏻‍♂️🙈

zackwine commented:
@PettitWesley Sorry I should have squashed some of those.  I wanted to give you credit for your work on this. Should I redo the PR?

hossain-rayhan commented on fluent-bit-kinesis.go:
```diff
@@ -83,7 +95,36 @@ func newKinesisOutput(ctx unsafe.Pointer, pluginID int) (*kinesis.OutputPlugin,
+			logrus.Errorf("[kinesis %d] Invalid 'concurrency' value %s specified: %v", pluginID, concurrency, err)
+			return nil, err
+		}
+		if concurrencyInt < 0 {
```
Q: Do we need to check any upper limit? Like, any value greater than 10 would be considered as invalid.

zackwine replied:
Good point, I'll add an upper limit of 10.  Given there is only one thread (fluentbit coroutine) feeding the output that is more than enough concurrency.

hossain-rayhan commented on kinesis/kinesis.go:
```diff
@@ -331,7 +425,7 @@ func (outputPlugin *OutputPlugin) randomString() string {
 // getPartitionKey returns the value for a given valid key
 // if the given key is emapty or invalid, it returns a random string
-func (outputPlugin *OutputPlugin) getPartitionKey(record map[interface{}]interface{}) string {
+func (outputPlugin *OutputPlugin) getPartitionKey(records []*kinesis.PutRecordsRequestEntry, record map[interface{}]interface{}) string {
```
Why do we need this new parameter `records`?

PettitWesley replied:
This function uses the record list to set the index. Previously, the record slice was stored in the plugin structure. That does not work if the plugin is concurrent though- each goroutine needs its own slice to store records.

hossain-rayhan replied:
Maybe I am wrong. But, I can see, you already removed the line(349) where we used `records`. I can't see the usage of your new `records` parameter.

zackwine replied:
Yes, I'll remove the records parameter.  The input record is the only required parameter now.

PettitWesley commented:
> Should I redo the PR?

@zackwine No... we can squash this change to one commit (with you as author) in the GitHub UI when it's ready to merge. For now I wouldn't worry about it.

zackwine committed:
Address code review items.

Added upper limit of 10 to concurrency.
Removed the unused records parameter from getPartitionKey.
Fixed improved error handling of configuration parameters.

Signed-off-by: Zack Wine <zwine@synamedia.com>

hossain-rayhan commented on README.md:
```diff
@@ -16,7 +16,9 @@ If you think you’ve found a potential security issue, please do not post it in
 * `time_key`: Add the timestamp to the record under this key. By default the timestamp from Fluent Bit will not be added to records sent to Kinesis.
-* `time_key_format`: [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html) compliant format string for the timestamp; for example, `%Y-%m-%dT%H:%M:%S%z`. This option is used with `time_key`. 
+* `time_key_format`: [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html) compliant format string for the timestamp; for example, `%Y-%m-%dT%H:%M:%S%z`. This option is used with `time_key`.
+* `concurrency`: Specify a limit of concurrent go routines for flushing records to kinesis.  By default concurrency is set to 0 and records are flushed in Fluent Bit's single thread. This means that requests to Kinesis will block the execution of Fluent Bit as whole.  If this value is set to `4` for example then calls to Flush records from fluentbit will spawn concurrent go routines until the lim ...
```
Add a line describing the default upper limit (10) of `concurrency`.

hossain-rayhan commented on kinesis/kinesis.go:
```diff
@@ -75,26 +77,28 @@ type OutputPlugin struct {
+	PluginID      int
+	random        *random
+	Concurrency   int
+	retryLimit    int
```
nit: If I am not wrong, this is `concurrency_retries`. Can we rename it to `concurrencyRetryLimit` to be consistent.

zackwine committed:
Address code review items (2).

Rename concurrency retires variable to concurrencyRetryLimit to be more consistent.
Update docs to mention concurrency option maximum value (10).

Signed-off-by: Zack Wine <zwine@synamedia.com>

zackwine commented:
Retesting with the tip of fluentbit (as of June 22nd), and this branch shows the throughput remains stable at ~3100 records/second.  This is up from ~2100 recoreds/second when I first opened the ticket:  https://github.com/fluent/fluent-bit/issues/2159

More on the test the produced the numbers above:

- Our fluentbit is configured as Kubernetes daemonset forwarding logs from containers
- The test runs 4 containers sending simulated application logs (averaging about 1K in size).  The rates above are the sum of all 4 containers.
- There are fairly intensive lua filters configured in fluentbit for parsing/routing our logs
- The test runs for a period longer than 10 mins at the rates mentioned above, with the expectation that 100% of the logs are sent to kinesis

hossain-rayhan commented:
@zackwine This is huge and thanks for your contribution and the test update.

hossain-rayhan commented on kinesis/kinesis.go:
```diff
@@ -194,36 +197,120 @@ func (outputPlugin *OutputPlugin) AddRecord(record map[interface{}]interface{},
 // Returns FLB_OK, FLB_RETRY, FLB_ERROR
-func (outputPlugin *OutputPlugin) Flush() int {
-	retCode, err := outputPlugin.sendCurrentBatch()
+func (outputPlugin *OutputPlugin) Flush(records *[]*kinesis.PutRecordsRequestEntry) int {
```
nit: I was confused about the parameter type `*[]*kinesis.PutRecordRequestEntry`.  Ignore my knowledge with golang syntax. I think you are correct here.

hossain-rayhan reviewed (approved):
I didn't run any test. The code for implementing currency and retry mechanism looks good to me. Ship! Great work.

zackwine commented:
@PettitWesley do you need to review this before merging?  What is the process?  Are there more tests that need to be executed?

PettitWesley commented:
@zackwine Someone should run the integ tests as described here: https://github.com/aws/aws-for-fluent-bit#developing-features-in-the-aws-plugins

Also... github says their are conflicts with mainline; please fix those.

@sonofachamp Can you do final review/merge once Zack rebases with mainline.

zackwine commented:
@PettitWesley 

I ran the integ test for kinesis:

```
make integ-kinesis-dev
...
Recreating test_kinesis_validate-s3_1 ... done
Attaching to test_kinesis_validate-s3_1
validate-s3_1  | [TEST SUCCESSFULL] Found all the log records.
test_kinesis_validate-s3_1 exited with code 0
```

I also attempted to merge master into my branch, but didn't find any new commits.

sonofachamp commented on kinesis/kinesis.go:
```diff
@@ -194,36 +197,120 @@ func (outputPlugin *OutputPlugin) AddRecord(record map[interface{}]interface{},
+		return output.FLB_RETRY
+	}
+
+	go outputPlugin.FlushWithRetries(count, records)
+
+	return output.FLB_OK
```
I have a concern with these changes that we're trading off reliability for performance.

My understanding is when we return an `FLB_OK` to the Fluent Bit engine we're acknowledging that the chunk of data Fluent Bit engine handed off was successfully processed. Fluent Bit engine would no longer be able to provide the built-in retry mechanism with that data as it assumes everything has been processed. With these concurrency changes we're hacking around that feature by returning `FLB_OK` earlier than we've processed the data. If concurrency retries are exhausted we're unable to tell the Fluent Bit engine `FLB_RETRY` and we've lost those records.

PettitWesley replied:
> I have a concern with these changes that we're trading off reliability for performance.

That's true, and that's why this new feature is opt-in.

zackwine replied:
That is a valid concern.  The default behavior is to not enable concurrency, so the Fluent Bit engine retry mechanism will continue to be utilized.

sonofachamp replied:
I think that's clear to us as we're directly involved in the changes, but for other customers and users I think it would be a surprise to opt in to any feature and end up with data loss. I don't agree with the reliability for performance tradeoff here or the hacky nature of returning `FLB_OK` early, but I can disagree and commit if others think this is the right thing to do. At a minimum we should log a warning message or find a better way to communicate that opting into the feature can result in data loss.

PettitWesley replied:
Makes sense. Let's explicitly call this out in the docs for this option? 

Zack built this feature because it is necessary for him to reach his throughput needs.  His testing has been pretty thorough and has shown that it does lead to a decent bump in sustained throughput.

sonofachamp replied:
@zackwine Would you be open to updating the exposed configuration options to include the `experimental_` prefix? I think this will make it clear that stability is not yet **guaranteed** and enable us to more easily pivot in the future if necessary. We'll also release this feature as experimental and update the docs to outline the potential cost associated with using these new fields.

zackwine replied:
@sonofachamp I'll make the changes to prepend `experimental_` to concurrency options.  And add a bit in the documentation about the additional resource usage, and potential for dropped records.

For our use case, the cost of not using the concurrency option is greater.  Enabling concurrency fixes the following 2 major issues for our use case:

- Repeated records when backpressure is received from kinesis (due to the retry logic in fluentbit not being aware a subset of the records were successfully posted).  With the concurrent retry logic if the Kinesis shard throughput is exceeded records will not be duplicated.
- Record throughput goes up about 70-90% with a concurrency value of 3.  This is particularly applicable if the source of records is a file that is rotated (read kubernetes docker logs) where a clock is ticking until the logs are rotated and lost at very high volumes.  Without the concurrency option enabled we will drop more records than with `concurrency=3` in my testing.

sonofachamp replied:
Thanks for some insight into your use case and results.

We really appreciate your contribution and I'm glad to hear these changes will improve your use case. While I understand for your use case the tradeoff is acceptable, it is not yet clear whether that tradeoff will be acceptable to all users of this feature in the future and the experimental additions will give us some time to get more data before locking it down.

Based on my understanding of the issues you were seeing I believe adding concurrency to the plugin is actually an indirect solution as I don't think the issues mentioned above are unique to only this plugin:
* Configurable options to control record counts or collective payload size could help solve the need to deal with partial record delivery in output plugins
* Concurrency is simply pushing that throughput threshold to missing file rotations a little bit further back

I think both of these should ideally be solved in the contract Fluent Bit provides between the engine and plugins. I'm not sure how feasible this actually is for the Go plugin interface, but hopefully this outlines why I think this concurrency solution is indirect.

PettitWesley replied:
> Record throughput goes up about 70-90% with a concurrency value of 3. This is particularly applicable if the source of records is a file that is rotated (read kubernetes docker logs) where a clock is ticking until the logs are rotated and lost at very high volumes. Without the concurrency option enabled we will drop more records than with concurrency=3 in my testing.

This is really impressive!!

To add to what Nick said though, we're a little bit uncertain about this change because it kind of fundamentally changes the contract between the plugin and Fluent Bit. Hence the ask to prefix all of the new options with `experimental_`. That leaves open the option to change things in the future, while allowing users to choose this feature after reading the docs and making an informed choice. 

Always returning FLB_OK will break fluent bit's i[nternal monitoring and exported Prometheus metrics](https://docs.fluentbit.io/manual/administration/monitoring)- please add a note in the readme that this feature breaks that as well.

And of course, thank you for all of this effort to try to improve the plugin for everyone!

PettitWesley commented:
@zackwine When I check out your code and try to rebase with master, there are changes.

zackwine committed:
Address code review.  Add data loss warning for concurrency.

PettitWesley commented on fluent-bit-kinesis.go:
```diff
@@ -66,6 +75,12 @@ func newKinesisOutput(ctx unsafe.Pointer, pluginID int) (*kinesis.OutputPlugin,
+	logrus.Infof("[firehose %d] plugin parameter concurrency = '%s'\n", pluginID, concurrency)
+	concurrencyRetries := output.FLBPluginConfigKey(ctx, "concurrency_retries")
+	logrus.Infof("[firehose %d] plugin parameter concurrency_retries = '%s'\n", pluginID, concurrencyRetries)
+	aggregated := output.FLBPluginConfigKey(ctx, "aggregated")
+	logrus.Infof("[firehose %d] plugin parameter aggregated = '%s'\n", pluginID, aggregated)
```
aggregated? I think this doesn't belong in this PR...

(Though if this means you're separately also working on KPL format, I am very happy 😃)

zackwine replied:
Yes.  Oops, I'll remove that.  I missed that change when switching contexts.

zackwine committed:
Address code review.  Rename the concurrency feature experimental.

zackwine committed:
Clean up imports.

sonofachamp commented:
The tests are failing:
```
→ make test
go test -timeout=120s -v -cover ./...
# github.com/aws/amazon-kinesis-streams-for-fluent-bit/kinesis [github.com/aws/amazon-kinesis-streams-for-fluent-bit/kinesis.test]
kinesis/kinesis_test.go:39:3: unknown field 'records' in struct literal of type OutputPlugin
kinesis/kinesis_test.go:42:3: unknown field 'lastInvalidPartitionKeyIndex' in struct literal of type OutputPlugin
kinesis/kinesis_test.go:77:35: not enough arguments in call to outputPlugin.AddRecord
        have (map[interface {}]interface {}, *time.Time)
        want (*[]*kinesis.PutRecordsRequestEntry, map[interface {}]interface {}, *time.Time)
kinesis/kinesis_test.go:79:28: outputPlugin.records undefined (type *OutputPlugin has no field or method records)
kinesis/kinesis_test.go:97:35: not enough arguments in call to outputPlugin.AddRecord
        have (map[interface {}]interface {}, *time.Time)
        want (*[]*kinesis.PutRecordsRequestEntry, map[interface {}]interface {}, *time.Time)
kinesis/kinesis_test.go:100:30: not enough arguments in call to outputPlugin.Flush
        have ()
        want (*[]*kinesis.PutRecordsRequestEntry)
?       github.com/aws/amazon-kinesis-streams-for-fluent-bit    [no test files]
FAIL    github.com/aws/amazon-kinesis-streams-for-fluent-bit/kinesis [build failed]
?       github.com/aws/amazon-kinesis-streams-for-fluent-bit/kinesis/mock_kinesis       [no test files]
FAIL
Makefile:37: recipe for target 'test' failed
make: *** [test] Error 2
```

zackwine committed:
Fix unit tests.
